docs/specs/tsd-001-deprecation.md

```diff
@@ -0,0 +1,181 @@
+plugin features including a timeline and minimal time-frames. Additionally, the
+specification defines a framework to annotate deprecations in the code and
+inform users about such deprecations.
+
+## User experience
+
+In the deprecation phase a warning will be shown at Telegraf startup with the
+following content
+
+```text
+Plugin "inputs.logparser" deprecated since version 1.15.0 and will be removed in 2.0.0: use 'inputs.tail' with 'grok' data format instead
+```
+
+Similar warnings will be shown when removing plugin options or option values.
+This provides users with time to replace the deprecated plugin in their
+configuration file.
+
+After the shown release (`v2.0.0` in this case) the warning will be promoted
+to an error preventing Telegraf from starting. The user now has to adapt the
+configuration file to start Telegraf.
+
+## Deprecation Process
+
+After reaching an agreement in the issue, you can start the deprecation process
+by following the steps below.
+
+### File issue
+
+In the filed issue you should outline which plugin, plugin option or feature
+you want to deprecate and *why*! Determine in which version the plugin should
+be removed.
+Consider moving the plugin to the Influx community repository as an external
+plugin if there is no straight-forward replacement.
+
+Try to reach an agreement in the issue before continuing and get a sign off
+from the maintainers!
+
+### Submit deprecation pull-request
+
+Send a pull request adding deprecation information to the code and update the
+plugin's `README.md` file. Depending on what you want to deprecate this
+comprises different locations and steps as detailed below.
+
+Once the deprecation pull-request is merged and Telegraf is released, we have
+to wait for the targeted Telegraf version for actually removing the code.
+
+You should consider writing a blog-post to make users aware of the scheduled
+removal of plugins, plugin options or option values.
+
+#### Deprecating a plugin
+
+When deprecating a plugin you need to add an entry to the `deprecation.go` file
+in the respective plugin category with the following format
+
+```golang
+    "<plugin name>": {
+        Since:     "<x.y.z format version of the next minor release>",
+        RemovalIn: "<x.y.z format version of the plugin removal>",
+        Notice:    "<user-facing hint e.g. on replacements>",
+    },
+```
+
+If you for example want to remove the `inputs.logparser` plugin you should add
+
+```golang
+    "logparser": {
+        Since:     "1.15.0",
+        RemovalIn: "1.40.0"
+        Notice:    "use 'inputs.tail' with 'grok' data format instead",
+    },
+```
+
+to `plugins/inputs/deprecations.go`. By doing this, Telegraf will show a
+deprecation warning to the user starting from version `1.15.0` including the
+`Notice` you provided. The plugin can then be remove in version `1.40.0`.
+
+Additionally, you should update the plugin's `README.md` stating the plugin is
+deprecated at the top (e.g. ***DEPRECATED***).
```

Comment:
Can you add an example, as I don't think you want the user to just put `***DEPRECATED***`, you want that + the notice to help guide the user.

Reply:
Again a copy and paste from the original document. For the now removed plugins that was exactly what was done. Should I change to what is done in `inputs.logparser` 

```
**Deprecated in Telegraf 1.15: Please use the [tail][] plugin along with the
[`grok` data format][grok parser]**
```
instead?

Reply:
you should feel free to update the specs as you put them up here

> Should I change to what is done in inputs.logparser

Yes I like your proposed change. We should ensure we guide the user with the call to action of what to move to next.

Reply:
I do like more how it is done in the aerospike plugin: https://github.com/influxdata/telegraf/blob/a8551659bc8ce34d13d37c58da31ebcb3f469d7e/plugins/inputs/aerospike/README.md?plain=1#L3-L5

I used it the same way to deprecate the ifname processor: https://github.com/influxdata/telegraf/pull/14899/files#diff-ad714a45b893f49a07ddd5132428edea0bd32db5650233832a200705738c0401R3-R4

Before the removal of deprecated plugins #13376, most of them were using this format:
```markdown
**Deprecated in version 1.30**: Please use [snmp_lookup][] plugin.
```

So should it be _version_ or _Telegraf_ in that section? To me te first looks better.

Reply:
Both versions would be possible according to this spec. I don't see why we would need to have `version of Telegraf` as all of this is version of Telegraf as we don't have any other version...

Reply:
Oh it was meant like a suggestion then? 👍

I didn't mean to put `version of Telegraf`, but `Deprecated in version 1.30` instead of `Deprecated in Telegraf 1.30`.